src/core/reactor.cc

```diff
@@ -1075,6 +1075,30 @@ void reactor::start_handling_signal() {
 
 namespace internal {
 
+signal_mutex::guard::~guard() {
+    if(_mutex == nullptr) {
+        return;
+    }
+    // Ensure the subsequent store isn't hoisted by the the
+    // compiler into the critical section it's intended to 
+    // protect.
+    std::atomic_signal_fence(std::memory_order_release);
+    _mutex->_mutex.store(false, std::memory_order_relaxed);
+}
+
+std::optional<signal_mutex::guard> signal_mutex::try_lock() {
+    if(!_mutex.load(std::memory_order_relaxed)) {
```

Comment:
Traditionally I'd expect a compare_exchange here. Are we saying we don't need it here because it's only really one thread that might enter this section twice (once form normal flow and then again from the inerrupt handler). So if we get interrupted after the load that is fine?

Reply:
Yeah, it's safe not to use a compare-exchange here for two reasons.  One the execution of the interrupt handle is mutually exclusive to the thread being interrupted. And two the interrupt handler will return the state of the mutex to how it was originally after completing. So if the interrupt occurs after a load the value of the load will still be correct after the interrupt finishes.

Reply:
The only real concern could've been partial writes if for some unlikely reason bools are written non-atomically. It's why I'm still wrapping the bool in `std::atomic`.